src/Gaffer/Animation.cpp

```diff
@@ -1026,6 +1083,50 @@ const Animation::Tangent& Animation::Key::tangent( const Animation::Direction di
 	return ( direction == Direction::In ) ? m_in : m_out;
 }
 
+Animation::TieMode Animation::Key::getTieMode() const
+{
+	return m_tieMode;
+}
+
+void Animation::Key::setTieMode( const Animation::TieMode tieMode )
+{
+	// check for no change
+
+	if( tieMode == m_tieMode )
+	{
+		return;
+	}
+
+	// make change via action
+
+	if( m_parent )
+	{
+		KeyPtr key = this;
+		TieMode previousTieMode = m_tieMode;
+		Action::enact(
+			m_parent,
+			// Do
+			[ key, tieMode ] {
+				key->m_tieMode = tieMode;
+				key->m_parent->m_keyTieModeChangedSignal( key->m_parent, key.get() );
+				key->m_parent->propagateDirtiness( key->m_parent->outPlug() );
+			},
+			// Undo
+			[ key, previousTieMode ] {
+				key->m_tieMode = previousTieMode;
+				key->m_parent->m_keyTieModeChangedSignal( key->m_parent, key.get() );
+				key->m_parent->propagateDirtiness( key->m_parent->outPlug() );
+			}
+		);
+	}
+	else
+	{
+		m_tieMode = tieMode;
+	}
+
+	ensureTieMode();
```

Comment:
How does undoing / redoing this work?  Do we not need to restore the previous tangent values when undoing, if they've been modified by ensureTieMode()?  And do we not need to apply ensureTieMode() again when redoing?

Reply:
any change to the tangant slopes is made by calling `setSlope()` which will perform the change via an action so it is recorded in the undo/redo system. However the scale ratio was not being recorded in the undo/redo system, that has now been fixed.